pkg/istio/authorizationpolicy/controller.go

```diff
@@ -0,0 +1,463 @@
+		},
+	})
+
+	adIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			item := Item{
+				Operation: model.EventAdd,
+				Resource:  obj,
+			}
+			c.ProcessConfigChange(item)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			item := Item{
+				Operation: model.EventUpdate,
+				Resource:  newObj,
+			}
+			c.ProcessConfigChange(item)
+		},
+	})
+
+	authzpolicyIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			item := Item{
+				Operation: model.EventAdd,
+				Resource:  obj,
+			}
+			c.ProcessConfigChange(item)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			item := Item{
+				Operation: model.EventUpdate,
+				Resource:  newObj,
+			}
+			c.ProcessConfigChange(item)
+		},
+		DeleteFunc: func(obj interface{}) {
+			item := Item{
+				Operation: model.EventDelete,
+				Resource:  obj,
+			}
+			c.ProcessConfigChange(item)
+		},
+	})
+
+	return c
+}
+
+// ProcessConfigChange is responsible for adding the key of the item to the queue
+func (c *Controller) ProcessConfigChange(item Item) {
+	c.Queue.Add(item)
+}
+
+// Run starts the main controller loop running sync at every poll interval.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.Queue.ShutDown()
+	wait.Until(c.runWorker, 0, stopCh)
+}
+
+// runWorker calls processNextItem to process events of the work queue
+func (c *Controller) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+// processNextItem takes an item off the queue and calls the controllers sync
+// function, handles the logic of re-queuing in case any errors occur
+func (c *Controller) processNextItem() bool {
+	itemRaw, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(itemRaw)
+	item, ok := itemRaw.(string)
+	if !ok {
+		log.Errorf("String cast failed for item %v", item)
+		c.Queue.Forget(item)
+		return true
+	}
+
+	log.Infof("Processing item: %s", item)
+	err := c.sync(itemRaw)
+	if err != nil {
+		log.Errorf("Error syncing k8s resource for item %s: %s", item, err)
+		if c.Queue.NumRequeues(itemRaw) < queueNumRetries {
+			log.Infof("Retrying item %s due to sync error", itemRaw)
+			c.Queue.AddRateLimited(itemRaw)
+			return true
+		}
+	}
+
+	return true
+}
+
+// sync is responsible for invoking the appropriate API operation on the model.Config resource
+func (c *Controller) sync(item interface{}) error {
+	castItem, ok := item.(Item)
+	if !ok {
+		return fmt.Errorf("Unable to cast interface")
+	}
+
+	// check if resource can be cast to service, athenzdomain, or authorizationpolicy object
+	_, svcCast := (castItem.Resource).(*corev1.Service)
+	_, adCast := (castItem.Resource).(*adv1.AthenzDomain)
+	_, apCast := (castItem.Resource).(*securityv1beta1.AuthorizationPolicy)
+	if !(svcCast || adCast || apCast) {
+		return fmt.Errorf("unable to cast interface to service or athenzDomain or authz policy object")
+	}
+
+	// dealing with service resource
+	switch true {
```

Comment:
I'm wondering if we should separate each of these event handlers into separate functions so that this function is just a bunch of function calls. It might make it easier to read. What do you think?